src/pages/Messages.jsx

```diff
@@ -6,34 +6,64 @@ import {
 import React, { useEffect, useState } from "react";
 import MessageBoard from "../components/MessageBoard";
 
-import { firestore } from "../firebase";
+import { auth, firestore } from "../firebase";
 import {
   collection,
   query,
   getDocs,
-  getDoc,
+  where,
   doc,
-  where
+  getDoc
 } from "firebase/firestore";
 
 const Messages = () => {
-  const [connectedChats, setConnectedChats] = useState([]);
   const [childId, setChildId] = useState("");
   const [connectedChatsObjects, setConnectedChatsObjects] = useState([]);
   useEffect(() => {
     const fetchData = async () => {
       try {
-        const collectionRef = collection(firestore, "Child");
-        // get users email from auth ===> auth.email
-        const q = query(collectionRef, where("email", '==', 'penpalprogram.murphycharity@gmail.com'))
-        const docRef = await getDocs(q)
-        if (!docRef.empty) {
-          const connectedChatsRef = docRef.docs[0].data()
-          setChildId(docRef.docs[0].id)
-          console.log("docs 0 ", docRef.docs[0])
-          connectedChatsRef?.connected_chats.forEach(chat => {
-            setConnectedChats([...connectedChats, chat])
-          })
+        if (!auth.currentUser?.uid) {
+          console.warn("error loading auth")
+          setTimeout(() => {
+            fetchData()
+          }, 2000)
+          return
+        }
+        const userDocRef = doc(collection(firestore, "users"), auth.currentUser.uid);
+        const userDocSnapshot = await getDoc(userDocRef);
+
+        if (userDocSnapshot.exists()) {
+          const userData = userDocSnapshot.data();
+          const connectedChatsRef = userData?.connected_penpals || [];
+
+          const connectedChatIds = connectedChatsRef.map(chat => chat.id);
+
+          const letterboxQuery = query(collection(firestore, "letterbox"), where("members", "array-contains", userDocRef));
+          const letterboxQuerySnapshot = await getDocs(letterboxQuery);
+
+          const messages = [];
+
+          for (const doc of letterboxQuerySnapshot.docs) {
+            const letterboxData = doc.data();
+            const lettersCollectionRef = collection(doc.ref, "letters");
+            const lettersQuerySnapshot = await getDocs(lettersCollectionRef);
+            const queryDocumentSnapshots = lettersQuerySnapshot.docs
+            const latestMessage = queryDocumentSnapshots[0].data()
```

Comment:
Please only fetch the data that you need (let database do the filtering), e.g.,
```suggestion
            const lettersCollectionRef = collection(doc.ref, "letters");
            const queryDocumentSnapshots = getDocs(query(lettersCollectionRef, where("deleted_at", "==", null), where("moderation.approved", "==", true), orderBy("deleted_at"), orderBy("moderation.approved"), orderBy("created_at", "desc"), limit(1)))
            const latestMessage = queryDocumentSnapshots[0].data()
```

Reply:
Please also make it fail safe in case the latest message does not exist or if it's attachments only.

Reply:
This also needs to be united with the results of unapproved moderations whereas the current user is the author of the message. It's okay to show these messages that are pending approval after the ones above.

Reply:
this filtering has changed since and we only retrieve messages we want correclty